REGRESSION (280917@main): Crash when omitting a visibility:hidden RenderLayer from the z-order tree.

mattwoodrow opened the pull request:
#### 0d9d9b7200189cbcf56cd043089e2ec6a7ac93fd
<pre>
REGRESSION (280917@main): Crash when omitting a visibility:hidden RenderLayer from the z-order tree.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277115">https://bugs.webkit.org/show_bug.cgi?id=277115</a>
&lt;<a href="https://rdar.apple.com/132358145">rdar://132358145</a>&gt;

Reviewed by Simon Fraser.

If we decide we can newly omit a RenderLayer in the z-order tree due to not
being visible, we need to also notify the compositor that these layers are
&apos;gone&apos; so that state can be updated correctly.

* LayoutTests/compositing/visibility/omitted-hidden-layers-crash-expected.txt: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-crash.html: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-inserted-expected.html: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-inserted.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateNormalFlowList):
(WebCore::RenderLayer::removeSelfAndDescendantsFromCompositor):
(WebCore::RenderLayer::setWasOmittedFromZOrderTree):
(WebCore::RenderLayer::collectLayers):
(WebCore::RenderLayer::updateLayerListsIfNeeded):
(WebCore::RenderLayer::updateDescendantDependentFlags):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: <a href="https://commits.webkit.org/281636@main">https://commits.webkit.org/281636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e79a73522018b4abecea7bceb0f2a07895dec06a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13083 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64439 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11059 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47548 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11284 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/7690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37144 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/52422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/29789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33835 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9655 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9968 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66173 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4454 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4473 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56496 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13449 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3696 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/compositing/visibility/omitted-hidden-layers-crash.html:
```diff
@@ -0,0 +1,27 @@
+<script>
+
+if (window.testRunner)
+    testRunner.waitUntilDone();
```
and `.dumpAsText()`

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -842,12 +866,16 @@ void RenderLayer::collectLayers(std::unique_ptr<Vector<RenderLayer*>>& positiveZ
+                layerList = makeUnique<Vector<RenderLayer*>>();
+            layerList->append(this);
+            accumulatedDirtyFlags.add(m_compositingDirtyBits);
+            setWasIncludedInZOrderTree();
```
It's a bit odd to set/clear these bits here. I'm not convinced that they are always updated correctly in the face of z-order tree mutations. I'd expect them to be dirtied via something closer to `m_visibleContentStatusDirty`.

mattwoodrow replied:
I'm less convinced of that. The logic for whether we include/omit layers is here, and takes into account more than just visibility state (the intrinsic compositing checks), so we'd need lots of dirtying all over the place otherwise.

mattwoodrow replied:
I guess the relevant dirty bit here is the existing z-order lists dirty state. That does already get set when we make something visible, though I think it's possibly not set in some reparenting cases.

That's an existing bug that would affect shipping, but I can try make a test case and fix it too.

mattwoodrow replied:
Ok, got a test that had a failure, and updated to fix that too.

smfr commented on LayoutTests/compositing/visibility/omitted-hidden-layers-crash.html:
```diff
@@ -0,0 +1,29 @@
+  <div style="position: relative; width: 150px; height: 400px; background-color: blue"></div>
+</div>
+
+</body>
```
Need some words like "This test should not crash".

smfr commented on LayoutTests/compositing/visibility/omitted-hidden-layers-crash.html:
```diff
@@ -0,0 +1,29 @@
+<body>
+
+<div style="width: 200px; height: 200px; overflow: scroll; border: 1px black solid" id="scroll">
+  <div style="position: fixed; width: 20px; height: 20px; background: red"></div>
```
Prefer not to use red.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -831,23 +832,48 @@ void RenderLayer::rebuildZOrderLists(std::unique_ptr<Vector<RenderLayer*>>& posZ
+    removeSelfAndDescendantsFromCompositor();
 
+    if (compositor().hasContentCompositingLayers() && parent())
+        parent()->setDescendantsNeedCompositingRequirementsTraversal();
```
Is `parent()` enough here, or should we go up to the enclosing stacking context (`paintOrderParent()`)?

mattwoodrow replied:
This is the existing behaviour for `RenderLayer::removeChild`, so hopefully it's sufficient :)

mattwoodrow committed:
REGRESSION (280917@main): Crash when omitting a visibility:hidden RenderLayer from the z-order tree.
https://bugs.webkit.org/show_bug.cgi?id=277115
<rdar://132358145>

Reviewed by Simon Fraser.

If we decide we can newly omit a RenderLayer in the z-order tree due to not
being visible, we need to also notify the compositor that these layers are
'gone' so that state can be updated correctly.

* LayoutTests/compositing/visibility/omitted-hidden-layers-crash-expected.txt: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-crash.html: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-inserted-expected.html: Added.
* LayoutTests/compositing/visibility/omitted-hidden-layers-inserted.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateNormalFlowList):
(WebCore::RenderLayer::removeSelfAndDescendantsFromCompositor):
(WebCore::RenderLayer::setWasOmittedFromZOrderTree):
(WebCore::RenderLayer::collectLayers):
(WebCore::RenderLayer::updateLayerListsIfNeeded):
(WebCore::RenderLayer::updateDescendantDependentFlags):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: https://commits.webkit.org/281636@main
